src/shogun/lib/SGVector.cpp

```diff
@@ -108,47 +114,57 @@ SGVector<T>::~SGVector()
 
 template <class T>
 SGVector<T>::SGVector(EigenVectorXt& vec)
-: SGReferencedData(false), vector(vec.data()), vlen(vec.size())
+: SGReferencedData(false), vector(vec.data()), vlen(vec.size()), gpu_vector(NULL)
 {
 
 }
 
 template <class T>
 SGVector<T>::SGVector(EigenRowVectorXt& vec)
-: SGReferencedData(false), vector(vec.data()), vlen(vec.size())
+: SGReferencedData(false), vector(vec.data()), vlen(vec.size()), gpu_vector(NULL)
 {
 
 }
 
 template <class T>
 SGVector<T>::operator EigenVectorXtMap() const
 {
+	if (on_gpu())
```

Comment:
I would put this check into a method that you call everywhere.
Just say: You called a direct memory access method when data is in GPU memory.